### PR TITLE
Add condition operations to KfDef for reconcile semantics

### PR DESCRIPTION
**Pull request by gabrielwen**
Adds functions to update and read conditions so that reconcile function could read from it and determine whether to terminate or retry.

Relates to #4056 
Blocking #4141

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/4166)

### CODE REVIEW

**Comment by gabrielwen** [comment]
@jlewi @yanniszark do we need to check duplicated conditions, e.g. plugin condition to a plugin?

File: bootstrap/pkg/apis/apps/kfdef/v1beta1/application_types.go [hunk 1]
```diff
@@ -160,6 +167,8 @@ type KfDefCondition struct {
 	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
 	// Last time the condition transitioned from one status to another.
 	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+	// Name of the component this condition points to.
+	Name string `json:"name,omitempty"`
```
Comment (yanniszark) [inline bootstrap/pkg/apis/apps/kfdef/v1beta1/application_types.go]: Should we use Reason instead of introducing a non-standard field?

**Comment by jlewi** [comment]
@gabrielwen either way. coordinator can get a list of the plugins from the KFDef spec. So it will know the kind for any plugin listed in the KFDef spec. So coordinator.go could programmatically look for appropriate conditions based on the plugin kinds.

So I don't think we need to hard code the conditions.

**Comment by gabrielwen** [comment]
@jlewi @yanniszark could you take another look?  this PR should be ready.

**Comment by jlewi** [comment]
/hold
/lgtm
/approve

@yanniszark any additional comments?

**Comment by jlewi** [comment]
/lgtm
/approve

@gabrielwen you can go ahead and remove the hold unless you want to wait for @yanniszark

**Pull request merged**